addons/website_sale/models/website.py

```diff
@@ -198,7 +198,11 @@ def get_current_pricelist(self):
         return pl
 
     def sale_product_domain(self):
-        return [("sale_ok", "=", True)] + self.get_current_website().website_domain()
+        if self.env.user.has_group('base.group_portal') or self.env.user.has_group('base.group_public'):
+            return [("sale_ok", "=", True), ("is_published", "=", True)] + self.get_current_website().website_domain()
+        else:
+            return [("sale_ok", "=", True)] + self.get_current_website().website_domain()
```

Comment:
Most of this is duplicated code, just extract ` [("sale_ok", "=", True)] + self.get_current_website().website_domain()` before the condition, and just add `("is_published", "=", True)` in the `if` condition.
Eg:
```py
domain = [..]
if condition:
    domain = AND(..)
return domain
```
No duplicate code, easier to read, no need of `else`, shorter line